vip-jetpack/connection-pilot/class-jetpack-connection-status-check.php

```diff
@@ -0,0 +1,200 @@
+<?php
+
+/**
+ * The Status Check is responsible for checking on the JP connection and handling the various disconnected scenarios.
+ * It stores a list notifications/errors that can be retrieved by the calling class and utilized however needed.
+ */
+class WPCOM_VIP_Jetpack_Connection_Status_Check {
+
+	/**
+	 * The option name used for keeping track of successful connection checks.
+	 */
+	const HEALTHCHECK_OPTION_NAME = 'vip_jetpack_connection_pilot_healthcheck';
+
+	/**
+	 * The healtcheck option's current data.
+	 *
+	 * Example: [ 'site_url' => 'https://example.go-vip.co', 'cache_site_id' => 1234, 'last_healthcheck' => 1555124370 ]
+	 *
+	 * @var mixed False if doesn't exist, else an array with the data shown above.
+	 */
+	private $healthcheck_option;
+
+	/**
+	 * The current site_url.
+	 *
+	 * @var string Site url.
+	 */
+	private $site_url;
+
+	/**
+	 * An error object describing the connection issue,
+	 * or why a (re)connection attempt failed.
+	 *
+	 * @var WP_Error A WP_Error object.
+	 */
+	private $connection_error;
+
+	/**
+	 * An list of notifications that need to be sent out by the Pilot.
+	 *
+	 * @var array Pilot notifications.
+	 */
+	public $pilot_notifications = array();
+
+	/**
+	 * Class constructor.
+	 * Populates the class properties and sets a fallback error.
+	 */
+	public function __construct() {
+		$this->healthcheck_option = get_option( self::HEALTHCHECK_OPTION_NAME );
+		$this->site_url           = get_site_url();
+		$this->connection_error   = new WP_Error( 'jp-cxn-pilot-default-error', 'The Jetpack connection is in an unknown state.' );
+	}
+
+	/**
+	 * The main method, used to start the status check.
+	 *
+	 * Runs connection checks and handles any problems.
+	 */
+	public function launch() {
+		$is_connected = WPCOM_VIP_Jetpack_Connection_Controls::jetpack_is_connected();
+
+		if ( true === $is_connected ) {
+			// Everything checks out. Update the healthcheck option and move on.
+			return update_option( self::HEALTHCHECK_OPTION_NAME, array(
+				'site_url'         => $this->site_url,
+				'cache_site_id'    => (int) Jetpack_Options::get_option( 'id' ),
+				'last_healthcheck' => time(),
+			), false );
+		}
+
+		if ( is_wp_error( $is_connected ) ) {
+			$this->connection_error = $is_connected;
+		}
+
+		// The connection check failed, let's handle the problem the best we can.
+		$this->handle_connection_issues();
+	}
+
+	/**
+	 * Handle connection issues. (Re)connect when possible, else add a notification.
+	 */
+	private function handle_connection_issues() {
+		if ( in_array( $this->connection_error->get_error_code(), array( 'jp-cxn-pilot-missing-constants', 'jp-cxn-pilot-development-mode' ), true ) ) {
+			return $this->notify_pilot( 'Jetpack cannot currently be connected on this site.' );
```

Comment:
It would be nice for the errors to include specific the "why" and what to do about it (otherwise we end up wasting time trying to figure that out.

Reply:
Every slack/irc message will also include the WP_Error that occurred when checking the connection status. So these will also help provide context.

Ideally, we'll be auto-resolving issues where possible as we start to see them occurring in the wild. If that's not possible though, then updating the messaging / adding a list of errors in the FG would be the next step.